File: Stack_Queue/LongestValidParenthesis.java
public class Solution {
    public int longestValidParentheses(String A) {
        Deque<Integer> st = new ArrayDeque<>();
        st.push(-1);
        int n = A.length();
        int maxlen = 0;
        for(int i=0; i<n; i++){
            if(A.charAt(i)=='('){
                st.push(i);
            }
            else{
                st.pop();
                if(!st.isEmpty())
                    maxlen = Math.max(maxlen, i - st.peek());
                else
                    st.push(i);
            }
        }
        return maxlen;
    }
}

File: Stack_Queue/LongestValidParenthesis.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        st = [-1]
        maxlen = 0
        for i in range(len(s)):
            if s[i] == '(':
                st.append(i)
            else:
                st.pop(-1)
                if not st:
                    st.append(i)
                else:
                    maxlen = max(maxlen, i-st[-1])
        return maxlen
